### PR TITLE
TRAFODION-1856: Revoke - object and column privilege checks not integrated for constraints

### PR DESCRIPTION
**Pull request by robertamarton**
Today, when revoking the object REFERENCES privilege, the revoke fails if there
are any RI constraints that require the privilege.  However, there may be column
level privileges that exist that would still allow the constraint to be present.
Conversely, when revoking column REFERENCES privilege, the revoke does not 
check to see if REFERENCES privilege has been granted at the object level.

### CODE REVIEW

File: core/sql/sqlcomp/CmpSeabaseDDLtable.cpp [hunk 1]
```diff
@@ -6309,6 +6304,23 @@ void CmpSeabaseDDL::alterSeabaseTableAddRIConstraint(
...
+
+       // If the user/role does not have REFERENCES privilege at the object 
+       // level, check to see if the user/role has been granted the privilege 
+       // on all affected columns
+       if (noObjPriv)
+         {
+           PrivMgrUserPrivs* privs = refdNaTable->getPrivInfo();
+           if (!privs->hasColReferencePriv(refdNAC->getPosition()))
+             {
+                *CmpCommon::diags() << DgSqlCode(-4481)
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/CmpSeabaseDDLtable.cpp]: Is there an enum for -4481?

File: core/sql/sqlcomp/PrivMgrMD.h [hunk 2]
```diff
@@ -56,35 +58,29 @@ typedef int             Lng32;
...
-    details += ", UID is ";
-    details += to_string((long long int) objectUID);
-    details += ", name is ";
-    details += objectName;
-    details += ", grantee is ";
-    details += to_string((long long int) granteeID);
-    details += ", is owner ";
-    details += (grantorIsSystem) ? "true " : "false ";
+    details = "column usage ";
+    details += " column number is ";
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/PrivMgrMD.h]: You'll get two spaces between "column usage" and "column number is". Is that what you want?

File: core/sql/sqlcomp/PrivMgrMD.h [hunk 3]
```diff
@@ -110,12 +106,35 @@ typedef struct {
...
+  : objectUID(0),
+    objectOwner(NA_UserIdDefault),
+    objectType(COM_UNKNOWN_OBJECT),
+    columnReferences(NULL),
+    updatedPrivs()
+  {}
+    
+  virtual ~ObjectReference ( void )
+  {
+    while(!columnReferences->empty())
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/PrivMgrMD.h]: This will abend if columnReferences is null.

File: core/sql/sqlcomp/PrivMgrMD.h [hunk 4]
```diff
@@ -110,12 +106,35 @@ typedef struct {
...
+      delete columnReferences->back(), columnReferences->pop_back();
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/PrivMgrMD.h]: Unusual coding technique (comma expression). Allows you to save some braces.

File: core/sql/sqlcomp/PrivMgrMD.h [hunk 5]
```diff
@@ -130,7 +149,70 @@ typedef struct {
...
+  PrivMgrDesc updatedPrivs;
+
+  ColumnReference * findColumn (int32_t columnOrdinal)
+  {
+    if (columnReferences == NULL)
+      return NULL;
+    for (int i = 0; i < columnReferences->size(); i++)
+    {
+      ColumnReference *pRef = (*columnReferences)[i];
+      if (pRef->columnOrdinal == columnOrdinal)
```
Comment (DaveBirdsall) [inline core/sql/sqlcomp/PrivMgrMD.h]: Instead of using a <vector>, you could use a <map> and get direct lookup rather than serial search.

**Comment by DaveBirdsall** [comment]
Aside from one place where there might need to be a null pointer check, these code changes look good to me.

**Comment by sureshsubbiah** [comment]
Looks good to me Roberta. Thanks.

File: core/sql/sqlcomp/CmpSeabaseDDLtable.cpp (hunk 1 above)
Comment (robertamarton) [inline core/sql/sqlcomp/CmpSeabaseDDLtable.cpp]: No, there are no enums for this range that I could find.

File: core/sql/sqlcomp/PrivMgrMD.h (hunk 3 above)
Comment (robertamarton) [inline core/sql/sqlcomp/PrivMgrMD.h]: Will fix this.

File: core/sql/sqlcomp/PrivMgrMD.h (hunk 4 above)
Comment (robertamarton) [inline core/sql/sqlcomp/PrivMgrMD.h]: I was having issues getting this code to work and I searched the web for examples.  This was one suggestion and it worked.

File: core/sql/sqlcomp/PrivMgrMD.h (hunk 5 above)
Comment (robertamarton) [inline core/sql/sqlcomp/PrivMgrMD.h]: I will note this and look at it for a subsequent commit.

File: core/sql/sqlcomp/PrivMgrMD.h (hunk 4 above)
Comment (DaveBirdsall) [inline core/sql/sqlcomp/PrivMgrMD.h]: Maybe I'm missing something. Isn't "delete columnReferences->back(), columnReferences->pop_back()" equivalent to { delete columnReferences->back(); columnReferences->pop_back(); }?

File: core/sql/sqlcomp/PrivMgrMD.cpp [hunk 6]
```diff
@@ -962,29 +979,41 @@ PrivStatus PrivMgrMDAdmin::getReferencingTablesForConstraints (
...
   // Select all the constraints that are referenced by the table
-  // create_time is included to order by the oldest to newest
   std::string selectStmt = "select distinct o.object_uid, o.object_owner, o.object_type, o.create_time, ";
   selectStmt += "trim(o.catalog_name) || '.\"' || ";
-  selectStmt += "trim (o.schema_name) || '\".\"' ||";
-  selectStmt += "trim (o.object_name)|| '\"' from ";
+  selectStmt +=   "trim (o.schema_name) || '\".\"' ||";
+  selectStmt +=   "trim (o.object_name)|| '\"' ";
+  selectStmt += ", c.column_number "; 
+  selectStmt += "from " + uniqueRefConstraintsMDTable + std::string(", ");
```
Comment (selvaganesang) [inline core/sql/sqlcomp/PrivMgrMD.cpp]: Why is that ","  should be std::string?  I wonder if std::string will trigger memory allocation in the malloc heap space

Comment (robertamarton) [inline core/sql/sqlcomp/PrivMgrMD.cpp]: The "+" operator for the c++ string class does not allow char * to be appended. (At least I have been able to do so), but if it is converted to a string first, it works.  As far as how the malloc heap is used, probably depends on how the compiler generates the code.

**Comment by sandhyasun** [comment]
Looks ready to merge.

**Pull request merged**